### PR TITLE
Improve docu and tests, add rttr:FormatNumGFs

### PR DESCRIPTION
**Pull request by Flamefire**
This bundles some improvements to the documentation, tests and adds the requested rttr:FormatNumGFs from #1185 

- Add a `doc` folder for documentation
- Convert the Lua doc from the wiki to files and fix references
- Ensure consistent Markdown via linter
- Fix a minor issue where lua code is ran twice for the equality check which can make mocks fail
- Refactor Lua interfaces to work with a new ILocalGameState which avoids some hacks required for testing and is the base for the  new feature which would otherwise be untestable
- Replace usage of Boost features with std features and Rttr features with Boost features where they exist

### CODE REVIEW

File: libs/s25main/network/GameClient.cpp [hunk 1]
```diff
@@ -758,7 +758,7 @@ bool GameClient::OnGameMessage(const GameMessage_Chat& msg)
 {
     if(msg.destination == CD_SYSTEM)
     {
-        SystemChat(msg.text, msg.player);
+        SystemChat(msg.text, (msg.player < game->world_.GetNumPlayers()) ? msg.player : GetPlayerId());
```
Comment (Flow86) [inline libs/s25main/network/GameClient.cpp:761]: shouldnt this the "system-player"? or do I understand correctly that this is then then the local player?

cant we simply write `SystemChat(msg.text)` here?

File: libs/s25main/network/GameClient.cpp [hunk 2]
```diff
@@ -1577,18 +1572,20 @@ void GameClient::SkipGF(unsigned gf, GameWorldView& gwv)
...
 
-void GameClient::SystemChat(const std::string& text, unsigned char player)
+void GameClient::SystemChat(const std::string& text)
 {
-    if(!ci)
-        return;
-    if(player == 0xFF)
-        player = mainPlayer.playerId;
-    ci->CI_Chat(player, CD_SYSTEM, text);
+    SystemChat(text, mainPlayer.playerId);
```
Comment (Flow86) [inline libs/s25main/network/GameClient.cpp:1577]: whats the difference between `mainPlayer.playerId` and `GetPlayerId()`?

File: libs/s25main/network/GameClient.cpp (hunk 1 above)
Comment (Flamefire) [reply libs/s25main/network/GameClient.cpp:761]: We have 1 example were this is used: When a player saves the game (which will cause a lag) he sends a message to be displayed to all players. In that case `msg.player` is set to that player and must be used. However we can also just send a message which should be displayed as `SystemChat(msg.text)` which is equivalent to `SystemChat(msg.text, GetPlayerId());` and done by setting `msg.player=0xFF`.

Kinda unusual but that's what it is currently.

File: libs/s25main/network/GameClient.cpp (hunk 2 above)
Comment (Flamefire) [reply libs/s25main/network/GameClient.cpp:1577]: Nothing: `unsigned GetPlayerId() const override { return mainPlayer.playerId; }`

Comment (Flow86) [reply libs/s25main/network/GameClient.cpp:1577]: then we should probably use `GetPlayerId()` there too, so its clear which one is used

Comment (Flamefire) [reply libs/s25main/network/GameClient.cpp]: Done: 7d27ec83c

**Pull request merged**